agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp

```diff
@@ -3090,6 +3135,79 @@ bool backhaul_manager::handle_1905_autoconfiguration_response(ieee1905_1::CmduMe
     return true;
 }
 
+bool backhaul_manager::handle_1905_beacon_metrics_query(ieee1905_1::CmduMessageRx &cmdu_rx,
+                                                        const std::string &src_mac,
+                                                        Socket *&forward_to)
+{
+    LOG(INFO) << "now going to handle BEACON METRICS QUERY";
+
+    // extract the desired STA mac
+    auto tlvBeaconMetricsQuery = cmdu_rx.getClass<wfa_map::tlvBeaconMetricsQuery>();
+    if (!tlvBeaconMetricsQuery) {
+        LOG(ERROR) << "handle_1905_beacon_metrics_query should handle only tlvBeaconMetrics, but "
+                      "got something else: 0x"
+                   << std::hex << (uint16_t)cmdu_rx.getMessageType();
```

Comment:
> otherwise the handler wouldn't have been called

This assumptions comes from the knowledge of the developer, there is nothing in the code that supports it. We, as developers, know that other wise it wouldn't have been called, but reality shows that these are _exactly_ the places the code breaks. It happens for example when code starts to move around, reconstructed - then, these hidden assumptions about the state of the inputs are broken. 
In robust code each function is a _standalone_ unit, does not have any assumptions about the flow that led to it being called, it is self contained and validates its inputs before working with them.

Reply:
> This assumptions comes from the knowledge of the developer, there is nothing in the code that supports it.

However, this check is making assumptions that don't exist. This function doesn't care what the type of the CMDU is that gets passed to it. It only cares about the TLVs that are in the CMDU. The caller has decided that this particular handler is OK for the CMDU. By adding the check here, you're not checking an invariant, but you're rather forcing the caller to be one specific caller. If you really want to force that, don't make a function of it...

To make it concrete: maybe a future update will define a different CMDU type that can be handled in the same way. Or maybe we will define a vendor-specific message that can be handled the same way.

Bottom line: I agree with Vitaly, this check is not only redundant, it is wrong.

That said, I don't think it's important enough to bother with removing it.

Reply:
Ok, I think I understand the confusion:
I _check_ for the correct tlv because I _expect_ it and _use_ it later in the code.
when the expected tlv is not found, I do the _best I can_ to give the user a hint about what I found, which is the message's type.

In general about the structure of this code:
We create concrete types of messages based on IDs (kind of a factory: in:ID, out:object)
These IDs _may_ (and will) conflict with the concrete types in certain scenarios.
The root of the complexity is the factory itself - there is no need to create concrete types based on IDs - and the logic is as simple as: at development time, if you know the ID - you know the concrete type (otherwise the factory wouldn't know what to construct). Therefore the factory itself is just a redundant, complex, unnecessary phase in the way.

Our code dresses and strips tlv object all the time out of their IDs.